index.php

```diff
@@ -0,0 +1,418 @@
+
+        download_plugin_skeleton($recipe);
+
+    } else if (!empty($formdata['buttondownloadrecipe'])) {
+
+        $recipestring = tool_pluginskel\local\util\yaml::encode($recipe);
+        download_recipe($recipestring);
+
+    } else if (!empty($formdata['buttonshowrecipe'])) {
+
+        $data = array('recipe' => $recipe);
+        $mform2 = new tool_pluginskel_step2_form(null, $data);
+
+        echo $OUTPUT->header();
+        $mform2->display();
+        echo $OUTPUT->footer();
+
+    }
+
+} else if ($step == 2) {
+
+    // Reconstruct the form.
+    $recipestub = array('component' => $component);
+    $data = array('recipe' => $recipestub);
+    $mform2 = new tool_pluginskel_step2_form(null, $data);
+    $formdata = (array) $mform2->get_data();
+
+    $recipestring = $formdata['recipe'];
+
+    if (!empty($formdata['buttondownloadrecipe'])) {
+
+        download_recipe($recipestring);
+
+    } else if (!empty($formdata['buttondownloadskel'])) {
+
+        $recipe = tool_pluginskel\local\util\yaml::decode_string($recipestring);
+        download_plugin_skeleton($recipe);
+
+    } else if (!empty($formdata['buttonback'])) {
+
+        $recipe = tool_pluginskel\local\util\yaml::decode_string($recipestring);
+
+        $generalvars = tool_pluginskel\local\util\manager::get_general_variables();
+        $componentvars = tool_pluginskel\local\util\manager::get_component_variables($recipe['component']);
+        $featuresvars = tool_pluginskel\local\util\manager::get_features_variables();
+
+        $rootvars = array_merge($generalvars, $featuresvars);
+        $rootvarscount = get_array_variable_count_from_recipe($rootvars, $recipe);
+
+        list($componenttype, $componentname) = core_component::normalize_component($component);
+        $componentfeatures = $componenttype.'_features';
+        $componentvarscount = array();
+        if (!empty($recipe[$componentfeatures])) {
+            $componentvarscount = get_array_variable_count_from_recipe($componentvars,
+                                                                       $recipe[$componentfeatures],
+                                                                       $componentfeatures);
+        }
+
+        $data = array_merge($rootvarscount, $componentvarscount);
+        $data['recipe'] = $recipe;
+
+        $mform1 = new tool_pluginskel_step1_form(null, $data);
+        $PAGE->requires->js_call_amd('tool_pluginskel/addmore', 'addMore');
+
+        echo $OUTPUT->header();
+        $mform1->display();
+        echo $OUTPUT->footer();
+    }
+}
+
+/**
+ * Returns the number of values for each variable array by examining the recipe.
+ *
+ * @param string[] $templatevars The template variables.
+ * @param string[] $recipe
+ * @param string $countprefix The prefix for the variable that will hold the number of values for the variables.
+ * @return string[]
+ */
+function get_array_variable_count_from_recipe($templatevars, $recipe, $countprefix = '') {
```

Comment:
It is quite ugly and bad practise to have functions defined in the main action scripts like this. Introduce a helper class for the web interface, or at least move it to locallib.php or so.